v1.8 backports 2020-08-07

jrajahalme opened the pull request:
* #12567 -- Ignore collisions for named ports that are not actually used in an egress policy (@jrajahalme)
* #12688 -- Cleanup unused methods and fields in pkg/policy (@aanm)
 * #12801 -- fqdn/dnsproxy: set SO_REUSEPORT on listening socket (@tklauser)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 12567 12688 12801; do contrib/backporting/set-labels.py $pr done 1.8; done
```

jrajahalme committed:
ipcache: Keep port name conflicts

[ upstream commit ad322a1764657c89d0677208b6dfba75101b5d85 ]

Do not warn of port name conflicts when ports are collected, but keep
the conficting ports. This allows (any) warnings to be postponed until
the named ports are used in a policy.

policy.NamedPortsMap is renamed to policy.NamedPortMap, and a new
policy.NamedPortMultiMap is defined, as well as an interface
policy.NamedPortsMap that abstracts the difference for callers of
GetNamedPort().

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
logging: Add rate limiting and use it for named port logging

[ upstream commit a02432bc1d9c8b48fba22b49ab55fb648678596a ]

Rate limit the warning messages emitted to the logs when a port name
can not be (uniquely) mapped to a port number.

Rate limit is set to 1 log every 10 seconds with a burst of 3.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
ipcache: Maintain named ports on demand

[ upstream commit fbd904915d53b5f1efdc47c4f4a106f12d21a07e ]

Pass Endpoint (or a subset interface) to potential users of Endpoint's
NamedPortsMap so that instead of preparing the map in advance it can
be created on demand in case it is actually needed.

NamedPortsMap is an interface that is implemented by both NamedPortMap
or NamedPortMultiMap, first of which is built from the POD spec, and
the latter is a collection of named ports across all the PODs in the
cluster. When a named port is used in an ingress policy, it can be
translated by using the PODs own NamedPortMap. When a named port is
used in an egress policy the named port can be one of any remote PODs,
so the collection of all the named ports in the cluster is used as the
individual destination POD is not known at policy computation
time. This could be optimized to the scope of PODs sharing the same
set of labels as the destination security IDs are known when policy is
configured.

Currently same port name resolving to different port numbers is
flagged as a runtime error and traffic to this/these ports is not
allowed. To get around this the port names should be made unique so
that each name maps to the same port number and protocol, or
preferably the port names in the egress policies should be replaced
with the desired port numbers.

The maintenance of the collection of named ports across the cluster
described above is started only after first needed for computing the
bpf map entries from an egress policy that uses a port name instead
the number. This way the related overhead is avoided if named ports
are never used in egress policies.

e.k8sPorts member must be accessed with the Endpoint locked. The
normal policy update path already has the endpoint locked, but incremental
selection updates originate from the selector cache without
any Endpoint locks, and therefore we need both locked and unlocked
variant of the GetNamedPortsMap().

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Refactor named ports for centralized logging

[ upstream commit fd0932409e57ff3753de127b46d6a9a8815461d8 ]

Refactor named ports so that error logging can be done in one place.
This makes adding rate limiting in a following commit easier.

Remove the need for policy.NamedPortsMap interface.

Add unit testing.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aanm committed:
pkg/endpoint: remove unused functions

[ upstream commit 06261bc972fc77e6ab91fb394898df4f438210c5 ]

The functions Get[Ingress|Egress]PolicyEnabledLocked are not used
anywhere so we can clean them up.

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aanm committed:
pkg/policy: remove unused returned field

[ upstream commit 491c9273c46fe8b7cc0655f5e43eb54a9234fa05 ]

This field was only used in unit tests, to simplify the code we can
remove its return.

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

tklauser committed:
fqdn/dnsproxy: set SO_REUSEPORT on listening socket

[ upstream commit 3abd218c224ce489ea271c728ba64e967ba2e66b ]

Forward port of #12728

Once we start re-using the same port for the DNS proxy across
restarts (see #12794), it's useful to set the SO_REUSEPORT option on the
listening port. This given the proxy a better chance to re-bind() upon
restarts.

Updates #12794

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

tklauser reviewed (approved):
Thanks, LGTM for my change.

rolinh reviewed (approved):
Thanks!

tklauser committed:
proxy: fix test build in v1.8 branch

The respective tests were removed from master in commit 03777e0ce0f1
("proxy: Move Kafka proxy to Envoy Go extensions").

This fixes the following test build failure:

./kafka_test.go:112:2: cannot use &proxyUpdaterMock literal (type *proxyUpdaterMock) as type logger.EndpointUpdater in assignment:
	*proxyUpdaterMock does not implement logger.EndpointUpdater (missing GetNamedPortLocked method)
./mock_test.go:49:42: undefined: policy.NamedPortsMap
FAIL	github.com/cilium/cilium/pkg/proxy [build failed]

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>
